master/internal/kubernetes/queue.go

```diff
@@ -0,0 +1,215 @@
+//  resource creation and deletion. An example of this is when a large HP search is created and
+//  then killed moments later. By having requests be buffered, if a a deletion request arrives
+//  prior to the creation request being executed, the requestQueue detects this and skips the
+//  unnecessary creation / deletion.
+//
+//  The message protocol consists of `createKubernetesResources` and `deleteKubernetesResources`
+//  messages being sent to the requestQueue. If it forwards the request to a worker, the worker
+//  will send the original task handler a `resourceCreationFailed` or a `resourceDeletionFailed`
+//  if an error was encountered while creating / deleting the resources. If a deletion request
+//  arrives before the creation request had been sent to the worker, the `requestQueue` will
+//  notify the task handler of this by sending a `resourceCreationCancelled` message.
+//  requestProcessingWorkers notify the requestQueue that they are available to receive work
+//  by sending a `workerAvailable` message.
+
+type requestQueue struct {
+	podInterface       typedV1.PodInterface
+	configMapInterface typedV1.ConfigMapInterface
+
+	queue                    []*queuedResourceRequest
+	pendingResourceCreations map[*actor.Ref]*queuedResourceRequest
+	availableWorkers         []*actor.Ref
+}
+
+func newRequestQueue(
+	podInterface typedV1.PodInterface,
+	configMapInterface typedV1.ConfigMapInterface,
+) *requestQueue {
+	return &requestQueue{
+		podInterface:       podInterface,
+		configMapInterface: configMapInterface,
+
+		queue:                    make([]*queuedResourceRequest, 0),
+		pendingResourceCreations: make(map[*actor.Ref]*queuedResourceRequest),
+		availableWorkers:         make([]*actor.Ref, 0, numKubernetesWorkers),
+	}
+}
+
+func (r *requestQueue) Receive(ctx *actor.Context) error {
+	switch msg := ctx.Message().(type) {
+	case actor.PreStart:
+		for i := 0; i < numKubernetesWorkers; i++ {
+			newWorker, ok := ctx.ActorOf(
+				fmt.Sprintf("kubernetes-worker-%d", i),
+				&requestProcessingWorker{
+					podInterface:       r.podInterface,
+					configMapInterface: r.configMapInterface,
+				},
+			)
+			if !ok {
+				return errors.Errorf("%s actor already exists", newWorker.Address())
+			}
+		}
+
+	case createKubernetesResources:
+		if err := r.receiveCreateKubernetesResources(ctx, msg); err != nil {
+			return err
+		}
+
+	case deleteKubernetesResources:
+		r.receiveDeleteKubernetesResources(ctx, msg)
+
+	case workerAvailable:
+		r.receiveWorkerAvailable(ctx)
+
+	default:
+		ctx.Log().Errorf("unexpected message %T", msg)
+		return actor.ErrUnexpectedMessage(ctx)
+	}
+
+	return nil
+}
+
+func (r *requestQueue) receiveCreateKubernetesResources(
+	ctx *actor.Context,
+	msg createKubernetesResources,
+) error {
+	if _, requestAlreadyExists := r.pendingResourceCreations[msg.handler]; requestAlreadyExists {
+		return errors.Errorf(
```

Comment:
blocking: Probably shouldn't retrurn errors, which brings down the `requestQueue` actor because this might happen if users start an trial, stop it, and start again. We probably just warn out it.

Reply:
same as the previous one